### PR TITLE
Genome dependent constructors in func registry

### PR DESCRIPTION
**Pull request by jigold**
Builds on #2306

### CODE REVIEW

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -137,10 +139,10 @@ object FunctionRegistry {
     }
   }
 
-  def call(name: String, args: Seq[AST], argTypes: Seq[Type]): CM[Code[AnyRef]] = {
+  def call(name: String, args: Seq[AST], argTypes: Seq[Type], rtTypConcrete: Option[Type] = None): CM[Code[AnyRef]] = {
```
Comment (danking) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: I've always wanted this actually. 👍

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 2]
```diff
@@ -62,7 +62,7 @@ object FunctionRegistry {
     }
   }
 
-  private def lookup(name: String, typ: TypeTag): Err[Fun] = {
+  private def lookup(name: String, typ: TypeTag, rtTypConcrete: Option[Type] = None): Err[Fun] = {
 
     val matches = registry(name).flatMap { case (tt, f, _) =>
       tt.clear()
```
Comment (danking) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: I don't understand why you don't need to clear `f.retType` too?

File: src/test/scala/is/hail/variant/GenomeReferenceSuite.scala [hunk 3]
```diff
@@ -126,4 +126,23 @@ class GenomeReferenceSuite extends SparkSuite {
     kt5.typeCheck()
     assert(kt5.forall("v37a.isAutosomal() && !v37na.isAutosomal() && v38a.isAutosomal() && !v38na.isAutosomal()"))
   }
+
+  @Test def testConstructors() {
+    val kt = hc.importTable("src/test/resources/sampleAnnotations.tsv")
+    val ktann = kt.annotate("""v1 = Variant(GRCh38)("chrX:156030895:A:T"), v2 = Variant(GRCh37)("X:154931044:A:T"),
+    |v3 = Variant(GRCh37)("1", 3, "A", "T"), l1 = Locus(GRCh38)("1", 100), l2 = Locus(GRCh37)("1:100"),
+    |i1 = Interval(GRCh37)("1:5-10"), i2 = Interval(GRCh38)("chrX", 156030890, 156030895)""".stripMargin)
```
Comment (danking) [inline src/test/scala/is/hail/variant/GenomeReferenceSuite.scala]: If I create a Variant/Interval/Locus without a reference what type does it get? Let's add a test for that.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 4]
```diff
@@ -81,6 +81,8 @@ object FunctionRegistry {
             None
         }
 
+        rtTypConcrete.foreach(f.retType.unify(_))
```
Comment (danking) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: This ignores the return type of `unify`, in particular, if the unification fails (i.e. a method only works for `Variant(GRCh38)` and `rtTypConcrete` is a `GRCh37` variant) the lookup must exclude this `f`.

**Comment by jigold** [comment]
@danking I'll add the genome reference to the Python type in another PR. The is stacked on #2375.

**Comment by jigold** [comment]
@catoverdrive This is failing due to `cannot set missing field for required type !!Struct{gt:Int32,ad:Array[Int32],dp:Int32,gq:Int32,pl:Array[Int32]}` in FisherExactTestSuite. I don't think it has anything to do with the changes I made in this PR. I did rearrange the Type gen methods, but those look correct to me.

Could you please take a look when you get a chance?

File: python/hail/representation/genomeref.py [hunk 5]
```diff
@@ -185,6 +185,46 @@ def GRCh38(cls):
...
+             "yContigs": ["Y"],
+             "mtContigs": ["MT"],
+             "par": [{"start": {"contig": "X","position": 60001},"end": {"contig": "X","position": 2699521}},
+                     {"start": {"contig": "Y","position": 10001},"end": {"contig": "Y","position": 2649521}}]
+            }
+
+        **Notes**
+
+        `name` must be unique and not overlap with Hail's pre-instantiated references: "GRCh37" and "GRCh38". `source` is
+        intended to be a string representing the source of the JSON file contents.
```
Comment (danking) [inline python/hail/representation/genomeref.py]: What is `source` for?

Comment (jigold) [reply python/hail/representation/genomeref.py]: I deleted it. It's not required. I had it in the json files that are in the hail directory to remember where I got the data from.

File: src/test/resources/fake_ref_genome.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{"name": "my_reference_genome",
+ "source": "/my/source",
```
Comment (danking) [inline src/test/resources/fake_ref_genome.json]: Shall we remove this as well?

**Comment by danking** [comment]
@jigold this should pass after a rebase, I merged the fix for the bug you experienced.

**Pull request merged**